### PR TITLE
support strides options for slice op

### LINKED ISSUE #112: Support strides option for slice
Link to https://github.com/webmachinelearning/webnn/issues/375#issuecomment-2292466613

image

```idl
dictionary MLSliceOptions : MLOperatorOptions {
  sequence<[EnforceRange] unsigned long> strides;
};
MLOperand slice(
    MLOperand input,
    sequence<[EnforceRange] unsigned long> starts,
    sequence<[EnforceRange] unsigned long> sizes,
    optional MLSliceOptions options = {}
);
```

### PR DESCRIPTION
**Pull request by BruceDai**
This PR is to fix #112.

@fdwr @huningxin @shiyi9801 PTAL, thanks!

### CODE REVIEW

File: src/slice.js [hunk 1]
```diff
@@ -8,19 +8,27 @@ import {validateSliceParams} from './lib/validate-input.js';
...
  */
-export function slice(input, starts, sizes) {
+export function slice(input, starts, sizes, {strides} = {}) {
   validateSliceParams(...arguments);
-  const outputShape = sizes;
+  strides = strides ?? new Array(input.rank).fill(1);
+  const outputShape = input.shape.slice();
+
+  for (let dimensionIndex = 0; dimensionIndex < input.rank; ++dimensionIndex) {
+    outputShape[dimensionIndex] = Math.floor(sizes[dimensionIndex] / strides[dimensionIndex]) +
```
Comment (shiyi9801) [inline src/slice.js:20]: Here you can use `Math.ceil(sizes[dimensionIndex] / strides[dimensionIndex])`, it works the same as `Math.floor(sizes[dimensionIndex] / strides[dimensionIndex]) + Number(sizes[dimensionIndex] % strides[dimensionIndex] != 0)`.

File: src/slice.js [hunk 2]
```diff
@@ -8,19 +8,27 @@ import {validateSliceParams} from './lib/validate-input.js';
...
+      Number(sizes[dimensionIndex] % strides[dimensionIndex]);
```
Comment (shiyi9801) [inline src/slice.js:21]: `Number((sizes[dimensionIndex] % strides[dimensionIndex]) != 0)`?

Comment (BruceDai) [reply src/slice.js]: Good catch!  Thanks!

**Comment by BruceDai** [comment]
@shiyi9801 Thanks for your reviewing! I addressed your comment in new commit. please take another look! ☕

**Comment by shiyi9801** [comment]
LGTM, thanks!

**Review by huningxin — APPROVED** [review]
LGTM!

File: src/slice.js [hunk 3]
```diff
@@ -8,19 +8,26 @@ import {validateSliceParams} from './lib/validate-input.js';
...
   const output = new Tensor(outputShape);
   for (let outputIndex = 0; outputIndex < sizeOfShape(outputShape); ++outputIndex) {
-    const loc = output.locationFromIndex(outputIndex);
-    const selectedInputLoc = loc.slice();
-    for (let i = 0; i < loc.length; ++i) {
-      selectedInputLoc[i] = loc[i] + starts[i];
+    const outputLocation = output.locationFromIndex(outputIndex);
+    const selectedInputLocation = outputLocation.slice();
+    for (let i = 0; i < outputLocation.length; ++i) {
+      selectedInputLocation[i] = starts[i] + outputLocation[i] * strides[i];
```
Comment (fdwr) [inline src/slice.js:28]: Nice clearer names :).

**Review by fdwr — APPROVED** [review]
👍

**Review by fdwr — APPROVED** [review]
👍

**Pull request merged**